### PR TITLE
[auto-materialize] Faster time window partition validation

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

Two simple changes here, aimed at addressing: https://github.com/dagster-io/dagster/issues/17429

1. Remove the `_get_validated_time_window_for_partition_key `method, as this was only used in the context of the `has_partition_key` method, and neither the body of that function nor the caller actually accessed the end time for the validated time window. Instead, just compute the start time for the partition window (much faster in simple cases), and run it through the same validation steps.
2. When we call `.empty_subset().with_partition_keys()`, stick with the in-memory format of a list of partition key strings, rather than doing the expensive conversion process immediately.

I also removed the `is_valid_partition_key` method, as this is essentially just an outdated version of `has_partition_key` (and does not actually catch all the cases that `has_partition_key` does, e.g. it does not consider the end_date of a time partitions def)

### CODE REVIEW

File: python_modules/dagster/dagster/_core/definitions/time_window_partitions.py [hunk 1]
```diff
@@ -309,14 +287,22 @@ def __hash__(self):
...
-    def _time_window_for_partition_key(self, *, partition_key: str) -> TimeWindow:
+    def time_window_for_partition_key(self, partition_key: str) -> TimeWindow:
         partition_key_dt = pendulum.instance(
             datetime.strptime(partition_key, self.fmt), tz=self.timezone
         )
+        # for simple cases, no need to invoke iterator
+        if self.is_basic_hourly:
+            return TimeWindow(partition_key_dt, partition_key_dt.add(hours=1))
+        if self.is_basic_daily:
+            return TimeWindow(partition_key_dt, partition_key_dt.add(days=1))
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:298]: do we need to worry about timezone stuff in this case? cc @gibsondan

Comment (gibsondan) [reply python_modules/dagster/dagster/_core/definitions/time_window_partitions.py:298]: oof. This is the relevant bit that could come into play when adding a day https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_utils/schedules.py#L157-L168

Wouldn't be an issue in UTC or if the hour is 0

**Comment by OwenKephart** [comment]
@sryza @gibsondan updated this to not make any spicy logical changes, just cuts out an unnecessary call. In terms of raw validation speed, this is around a 7x improvement (went from 143s to validate 1M partitions to 20s)

**Comment by sryza** [comment]
@clairelin135 mind taking a look at this one?

**Review by clairelin135 — APPROVED** [review]
These changes are reasonable to me!

**Pull request merged**